src/Juvix/Core/IR/Typechecker.hs

```diff
@@ -23,67 +74,216 @@ typeTerm ∷
+          <> ", it is *i, as required. "
+      logOutput $
+        "Checking that the variable (V) type checked to the input annotation, which is just type checked to be of usage 0 and type *i. "
+      typeTerm p ii g varType ann
+      logOutput $
+        passed
+          <> "The variable (V) is "
+          <> show (fst ann)
+          <> " and type "
+          <> show (snd ann)
+          <> ", as required. Checking that the result (R), with x of type V in the context, type checked against the input annotation. "
+      ty ← evalTerm p varType [] -- V
       typeTerm
-        p -- checks resultType is of type Star i
+        p -- param
         (ii + 1)
-        ((Local ii, (pi, ty)) : g)
-        (substTerm 0 (Free (Local ii)) resultType)
-        ann
-    _ →
-      throw @"typecheckError" UniverseLevelMustMatch
+        ((Local ii, (SNat 0, ty)) : g) -- add x of type V, with zero usage to the context
+        (substTerm 0 (Free (Local ii)) resultType) -- R, with x in the context 
+        ann -- is of 0 usage and type *i
+      logOutput $
+        passed
+          <> "Result (R) has usage of is of type *i. "
+          <> typechecked t ann
+    _ → do
+      logOutput $
+        failed
+          <> "The annotation is not of type *, it is of type "
+          <> show (snd ann)
+      throw @"typecheckError" (ShouldBeStar (snd ann))
 -- primitive types are of type *0 with 0 usage (typing rule missing from lang ref?)
 typeTerm _ ii _g x@(PrimTy _) ann = do
   ty ← quote0 (snd ann)
-  unless
-    (SNat 0 == fst ann && ty == Star 0)
-    (throw @"typecheckError" (TypeMismatch ii x (SNat 0, VStar 0) ann))
--- Lam (introduction rule of dependent function type)
-typeTerm p ii g (Lam s) ann =
+  typeTermIntroLog x ann
+  logOutput
+    ("patterned matched to be a primitive type term. Checking that input annotation is of zero usage. ")
+  if (SNat 0 /= fst ann)
+    then
+      ( do
+          logOutput $
+            failed
+              <> "The input usage is "
+              <> show (fst ann)
+              <> ", which is not zero. "
+          throw @"typecheckError" (UsageMustBeZero)
+      )
+    else do
+      logOutput "Checking that input annotation is of type *0. "
+      if (ty /= Star 0)
+        then
+          ( do
+              logOutput $
+                failed
+                  <> "The input type is "
+                  <> show ty
+                  <> ", which is not * 0. "
+              throw @"typecheckError" (TypeMismatch ii x (SNat 0, VStar 0) ann)
+          )
+        else do
+          logOutput $ passed <> typechecked x ann
+          return ()
+-- Lam (introduction rule of dependent function type), requires Pi (formation rule of dependent function type)
+typeTerm p ii g t@(Lam m) ann = do
+  typeTermIntroLog t ann
+  logOutput
+    "patterned matched to be a Lam term. Checking that input annotation has sigma usage and is dependent function type (Pi). "
   case ann of
     (sig, VPi pi ty ty') → do
```

Comment:
`sig` should be 0 or 1 according to the current lang ref, but currently that is not checked.